### ISSUE TITLE
Simplify the ease of downloading individual items from the portal

### ISSUE THREAD

**Issue by fedorov** [ImagingDataCommons/IDC-WebApp#1186]
Based on my experience and discussions, users struggle a lot with the complexity of download from IDC, and I think even after addressing #1182 it will be suboptimal.

Here's an idea - can we do something like what is done in the GCP storage bucket interface, where up to some point individual items can be downloaded directly, and after crossing some threshold, user is prompted with the command to run to download specific items? And in the latter case, user can copy the exact command line, paste it, and have the data.

Something like this:

![image](https://github.com/ImagingDataCommons/IDC-WebApp/assets/313942/7b476686-b161-4674-bd94-d73cdf95aad5)

Dear disgruntled IDC users: @denbonte @lennyN95 @salujajustin - please add your comments and thoughts about this suggestion - in support, disagreement, or with alternatives that you think might work better?

**Comment by LennyN95** [comment]
This is already a big improvement, much more inviting to continue and signaling to the user that he's on the right track.
Ideally, we'd show a command that downloads sample data from the selection (e.g. the first 2) - would that be possible?

**Comment by LennyN95** [comment]
Oh, and as Dennis and I were just discussing, the manifest script that you showed us, why can't we have that here as well?
As I understand it, the current user interaction path looks like `log in -> save as new cohort -> get manifest -> open  manifest and run the command from the comment on the top line`. Perhaps it might help to add a shortcut, such as a `Download data from manifest` button. This button then opens a popup explaining the next steps (ideally checking where we're in the process, e.g. whether the user is already logged in or not, etc.)

**Comment by fedorov** [comment]
All good points - keep them coming!!

And how could I possibly forget @surajpaib!

**Comment by salujajustin** [comment]
Even for 1 file we may want to still want to have to command line interface to keep everything cohesive. You can have an option to download sample data as @LennyN95 mentioned in each case.

**Comment by surajpaib** [comment]
This is definitely a welcome addition.

@fedorov It would be great to have this on all levels - Collection, Case, Study, and Series. I've often wanted to download specific patients and also specific series individually. Depending on the granularity of the selection, different methods to offer downloads to the end users could be designed, as @LennyN95 mentioned, in addition to being able to download the manifest. 

I really like the style google cloud provides in its documentation for different methods
![image](https://github.com/ImagingDataCommons/IDC-WebApp/assets/10467804/e37f5635-87a6-4c07-8bd3-0ae5df0f77a6)

and I see that IDC also uses this style for the manifest download
![image](https://github.com/ImagingDataCommons/IDC-WebApp/assets/10467804/76e3a9b5-1933-484c-9e90-2079e03a0086)


Maybe something like this as well when the user clicks on the download button with a subset of options popping up based on what is selected (for example, selecting at the collection level won't offer a direct file download, but selecting at the series level will)

Happy to help test and brainstorm!

**Comment by fedorov** [comment]
> If you need to download many cohorts however, a super long command line option might be intimidating. We could provide them as a zip as @denbonte mentioned or maybe you have a temporary file with all the things that need to be downloaded. Either way, it would be ideal just to run one command that handles everything

@salujajustin what I propose here is a shortcut for quick access to specific items. If one wants to download a complex cohort, then they will need to create a manifest. It is still one command to download everything, you just need to pass the manifest.

**Comment by Gibbsdavidl** [comment]
Good discussion!  

There's a lot of progress with the version under development including taking those cohorts (that can now be created without saving or logging in) **and** getting a manifest that can be passed to s5cmd – one command to get all the images in the cohort. We expect this next version to be released relatively soon.

However, the current direction of discussion / research is focused on exploring how to refine the filter set which defines the cohort. With the current portal, the left-side defines the filter, and everything that's returned is shown on the right-side. Unfortunately, the checkboxes on the RHS, although they look like they're making selections, are functionally not. Instead they control which information about the selected cohort is currently viewed in the portal. Building the functionality to make selections on the right side that refines the filter/cohort is a major change to the portal implementation, and something that we want to approach with some caution. It's definitely something we are devoting a lot of effort towards.

With respect to downloading a single case, or study, or even series, I really like the idea of having a button on the right-hand side of each -row that would create a "mini-manifest" that could be passed to s5cmd – then just a single command to get a single collection, patient, study, series or even one instance, but we would want to design it to work in concert with cohort filter refinement using interactions on the right-hand side. Also having the portal manage the download means pulling all the bytes from the bucket through the portal web app, which is really not going to be efficient.. Note that with on-prem, the web app is probably pulling it off a local disk, NFS mount, or Storage Area Network, and so it is necessary to pull the bytes through the web server running the portal anyway.

**Comment by fedorov** [comment]
> then just a single command to get a single collection, patient, study, series or even one instance, but we would want to design it to work in concert with cohort filter refinement using interactions on the right-hand side.

I am not sure what "in concert" would mean, but I would argue that this feature can be complementary and independent. I think those are different enough. 

> Also having the portal manage the download means pulling all the bytes from the bucket through the portal web app, which is really not going to be efficient

The way I suggested it, there would be no interaction between the portal and the bucket. The portal provides the s5cmd command, and the user runs s5cmd command from the terminal on their computer. s5cmd will pull all the bytes - not the portal.

**Comment by pieper** [comment]
Thanks for all the brainstorming on this - I just wanted to add my vote here for making it easier to do quick downloads of interesting data at the individual study level.  Personally, I often need some data to test a tool (e.g. a Slicer module) and would love to be able to drill down in IDC to find a representative set of images and download them quickly.

I don't mind using s5cmd or gsutil for bigger downloads, but just getting a zipfile from the browser would be a lot nicer for small datasets.

BTW, a nice to have feature for the portal would be to tell you statistics about your current selection, such as the number of images and the size of the current selection.

**Comment by fedorov** [comment]
> a nice to have feature for the portal would be to tell you statistics about your current selection, such as the number of images and the size of the current selection.

This was added in the latest release!

<img width="587" alt="image" src="https://github.com/ImagingDataCommons/IDC-WebApp/assets/313942/db4ba816-9ed2-4e9d-95f1-d77ecaae2381">

**Comment by fedorov** [comment]
Per discussion today with the ISB team: creating a zip file on the fly would be very difficult to do. If we want to do it on the server, this would require significant infrastructure work to either maintain a dedicated VM for zipping, or significantly expand the configuration of the VM we are using right now. There would also be questions about how long the zipped file would need to be stored. We could easily provide links to the individual files, but since individual series contain dozens if not hundreds, it would be of limited utility to the users. We could pre-create zip files, but we discussed that earlier, and it would require storage, maintenance, versioning, costs management, and all the related considerations.

The feature to provide a command to download using s5cmd should be easy to add.

@pieper let me know your thoughts. We can also discuss this specific feature at the upcoming meeting.

**Comment by pieper** [comment]
My thought was that we would create the zip file on the client (in the browser), either by passing api tokens to the client javascript or by proxying the download stream through something lightweight on the server.  Once the data is collected it can be zipped and offered for download.  I have done this before so I know it's possible.  We can always impose a limit of like 2G and if there's more that that we tell people they need to use one of the other methods.

**Comment by fedorov** [comment]
We don't need any tokens, since the data is available from the buckets via unauthenticated URL. But @s-paquette told me browsers have very small memory limit to accommodate zip file creation. I added to the Tue meeting agenda, so let's also discuss there!

**Comment by pieper** [comment]
You should be able to do one or two gig in the browser easily enough and downloading a study or two that way would still be very valuable for people I think.  Getting a little data for inspection and then being told you need to use a different process if you like it and want bulk access seems like a reasonable workflow.

I suggested the token since I'm not sure if the gcp buckets support CORS and I'm not sure if you can use a google browser based javascript API to get public data, but if there's a way I think that leveraging the client side would be valuable.

I'll be out on vacation and will probably miss the calls the next two weeks.

**Comment by fedorov** [comment]
> I'm not sure if the gcp buckets support CORS and I'm not sure if you can use a google browser based javascript API to get public data

We can configure CORS on the bucket. I know because we had to white-list VolView app to do this, and it accesses the data directly from the bucket. So this should not be an issue.

@s-paquette can you share your thoughts on this?

**Comment by s-paquette** [comment]
@pieper We have some experience in dealing with loading large (>1GB) amounts of data into the browser in one of our other projects, and found the performance hit to be enormous and not particularly workable. For amounts under eg. a couple hundred megabytes, that runs fine, but anything over that and it turns into a significant problem for the user (this was us as developers testing on machines with 16+ GB of RAM; average users may or may not have that much). If that would be useful to users, we can try to come up with a way to do that for smaller sets, but it leads us into the next point.

The current web application and portal was not specced or designed to manage downloads of blocks of files from GCS to a user. Realistically it can't in its current form, due to the constraints of AppEngine Flex (no local disk storage, extremely expensive RAM, restrictive timeout from the load balancer). A different system would need to be put into place to address this kind of requirement. There's the additional issue of the storage of the files: in cloud storage, a study or a series is not a single object which can be requested via HTTPS. You must individually request each file. Depending on how many, that can easily run into problems with the web server waiting on GCS as it performs each individual request, then needing to package those for relay to the user. Introducing the portal as a middle-man will produce a not insignificant delay and incur costs for us in cross-loading the data to the web server, packaging it, and sending it on.

In the end, we would be in effect creating a user interface for easier use of GCS/AWS. That's a significant undertaking, and would require a fair amount of our time to scope out and produce.

**Comment by pieper** [comment]
Let's not forget that we are expecting people to have browsers that can display these images (uncompressed) so I don't think it's a stretch to think that they will be able to store enough of them in a compressed zip file to handle a lot of practical cases.  To me it would be incredibly logical to have a download button next to the eye button at the study level that tries to download the study from the bucket to make a zip in the browser.  If that operation fails, the error message can say that the user needs to use one of the other download methods.  All of this can happen in the client with no server involvement.

**Comment by s-paquette** [comment]
@pieper Unfortunately since the error would be an out of memory error, it's not something we can detect. The browser will simply crash. We're better off setting a specific cap on the study (or series) size we'll allow to be downloaded this way. We can titrate for that, but my expectation is, at most, we'd want to only allow about 1GB. I'm not sure how many studies fall under that, but it's easy enough to check them coming in and not enable the capability unless they do.

**Comment by pieper** [comment]
> Unfortunately since the error would be an out of memory error, it's not something we can detect. The browser will simply crash. 

Hmm, I'm not sure that's the case.  If we know how much data we plan to download we can try to allocate a buffer for it and if it fails the browser won't crash.  If it succeeds we can copy the results of successive downloads into the allocated buffer (or just add them to the zip to be compressed).

But I agree this method would have some practical limit so we shouldn't enable it for whole collections.  Just giving an easy way to get a study would be really handy for real-world imaging research work, so I think this is valuable enough to investigate a bit more.

**Comment by s-paquette** [comment]
@pieper There isn't, that I know of, a way to allocate memory--there's an API for checking what's available (both the JS heap and via the storage API), but that's an estimate so we'd need to build in some leeway. Since we'd be using the JS heap to zip the files that's arguably the more important one; the maximum for that in most browsers is 4G, so we'd be looking at any study with under about 2G of files (assuming the other 2G of JS heap is under use by other tabs, or the user simply doesn't have that much memory, etc.). A quick query of the study index gets me this:

```SELECT COUNT(*)
FROM `idc-dev-etl.idc_v15_dev.study_index`
WHERE instance_size/(1024*1024*1024) < 2;

135758
```

This is most of them; assuming we need to go under 1G, that drops to 130877 so still quite a few.

**Comment by fedorov** [comment]
Per discussion today, for the current release, we will be adding a "mini manifest" - s5cmd command that can be copied to download files at either series or study level (since some patients have as many as 1500 or ~800 series, those "mini manifests" may become too large, but we can add patient level later, based on user feedback, if needed).

In the subsequent release we will experiment with allowing to download individual series directly as a zip file (since as shown above, we should be able accommodate most of series).

**Comment by pgundluru** [comment]
![image](https://github.com/ImagingDataCommons/IDC-WebApp/assets/85526819/335b26b6-29ac-41be-b0f0-0a2639bcf7d5)

**Comment by pgundluru** [comment]
Mini- Manifests S5cmd files - that can be downloaded are now added to study and series level

**Comment by fedorov** [comment]
The mini-manifest feature was added back in #1239. The next step is to address direct download from the browser, which was discussed in https://github.com/ImagingDataCommons/IDC-WebApp/issues/1186#issuecomment-1654161202:

> In the subsequent release we will experiment with allowing to download individual series directly as a zip file (since as shown above, we should be able accommodate most of series).

**Comment by s-paquette** [comment]
Note that the download feature will be far too big to address in Release 41 (i.e. before the November holiday break) so I'm removing the milestone for now. We can address it in Release 42 or 43.

**Comment by pieper** [comment]
Following up on [my comment above](https://github.com/ImagingDataCommons/IDC-WebApp/issues/1186#issuecomment-1636413001), I wanted to say that the [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) supported by Chromium and Edge is super efficient for saving to files and could be overlapped with downloads to make a very easy to use and probably fast download option.

We make use of it in the [`dicom-curate`](https://github.com/bebbi/dicom-curate) project for reading/writing and it can be used in web workers very efficiently.

**Comment by fedorov** [comment]
@s-paquette I think we should seriously consider making this issue top priority!

**Comment by fedorov** [comment]
@pieper @s-paquette per discussion today, we need to identify an s3 library to fetch data directly from the buckets from the browser client. Steve do you have suggestions?

**Comment by s-paquette** [comment]
It looks like the best option is to bundle up the AWS SDK s3 library with s3-zip, which should then let us arbitrarily bundle whole buckets/folders/individual instances into zips. Note this is all done browser side, so the actual speed of it is wholly dependent on the user's own machine and connection.

**Comment by fedorov** [comment]
I think it is not obvious decision to zip things up. Why not stream from the bucket to disk file hierarchy? This is what idc-index does, and this is also what TCIA download client does.

**Comment by pieper** [comment]
I don't see the need for any special libraries, at least not on chrome/edge and probably not firefox or safari either, although I haven't tried on those platforms.  I believe you can access the buckets directly with xmlhttprequests (i.e. just as plain unauthenticated GET requests) and then you can save them in a directory hierarchy at a location of the user's choice.  You can use web workers to handle each download and file save so that all the I/O is overlapped.  No zipping required or desired in my opinion.

Even if it turns out you can only do this on chrome/edge, I think that would be a big enough slice of the user base to go ahead with it.  Other browsers could fall back to the s5cmd/idc-index approach, which I think is about as good as the zip approach, and easier since it already exists.

I'm tied up for the next few weeks, but then I'm happy to make a prototype so we can test out my hypotheses before investing more time in it (or someone else can do it in the meantime).

**Comment by s-paquette** [comment]
IIRC all current modern browsers support the File System API. There could be variance in their interfaces to it, but it's probably not much. The main benefit of the AWS library (which doesn't need to include Zip if we don't want it too, I just recall someone--maybe David Clunie?--indicating that might be preferable for larger numbers of files. 

The benefit of the AWS JS library is we're not needing to write and maintain large blocks of JS code ourselves; that is done by Amazon and it's specifically written to take advantage of any AWS aspects of S3. We certainly can write something of our own, but it might take longer and require more maintenance. Those are the basic trade offs; managed by Amazon, or managed by us.

**Comment by fedorov** [comment]
Since @pieper offered to make a prototype, and he has experience doing similar things in another project, I suggest we take him up on this offer.

**Comment by fedorov** [comment]
> Also it's not a big deal to detect browsers where it doesn't work and just not offer the option or leave it disabled with a tooltip. I wouldn't worry about a workaround.

I agree! We can consider investing effort into workarounds and other browsers if we hear this is a real pain point from the users.

**Comment by s-paquette** [comment]
The issue isn't the File System API but the File System Access API. The former allows downloading the files to the browser but the later--which is what includes the directory picker--is needed to save the files to the wider file system. As a result although webkit and mozilla can download the files, they can't move them to the file system via JS alone.

The solution for these browsers is to load the file(s) into the browser, then trigger a download as if the user had clicked on a download link. This is how we do the current manifest option, so it's accessible to all browsers. 

For something larger and/or with multiple files like we've proposed here, that won't work; we really need to use web workers and the File System Access API, but that's only available in Chromium browsers. it's not difficult to detect the user agent and disable, however.

**Comment by s-paquette** [comment]
@pieper The issue with that option is the input element is for uploading files to the server, not downloading them. We won't be able to use the directory chosen to actually save out the file with Firefox or Safari, as only the File System Access API can do that, which they both lack. (More specifically, Safari has it, but only for local browser storage, not for the broader file system like Chrome),

**Comment by fedorov** [comment]
To clarify Steve's question: focusing only on Chrome and forgetting all other browsers for now, will you be able to come up with a prototype @s-paquette to pull files directly from the bucket and save them to a local directory selected by the user or should we ask Steve ?

**Comment by s-paquette** [comment]
@fedorov @pieper Perfectly happy for Steve to draw up a prototype--note that we don't use NodeJS or React so it does need to be pure JS and jQuery

**Comment by s-paquette** [comment]
@pieper @fedorov Alright, looking like I spoke too soon re: instance IDs (i.e. full paths for a given series). This is the issue we ran into with trying to store all of the paths for a Study or Series in the first place; it makes the Solr datasets gigantic, due to the sheer amount of text involved (even just adding SOPInstanceUIDs to a series added over 2GB to the size of the series dataset, from 800MB to 3.5GB; studies would probably grow a similar amount). We can definitely still do this, but we do need to consider how this will impact the responsiveness of the WebApp in general, since a larger Solr index means more time gathering response information. 

One option is to add an additional Solr dastaset which is just the complete paths to the instances for a given study or series (and if we want, case or collection) and nothing else. When a user clicks on download, it's the endpoint for this dataset which is called to fetch those paths out. This would be a relatively light lift and keep the main dataset from being overloaded with information unrelated to its primary job (letting a user search for and enumerate collections/patients/studies/series and generate manifests listing them).

**Comment by pieper** [comment]
Thanks for the extra info @s-paquette, that does make sense.  You may have seen on slack we've been looking into this and there's also the CORS issue on the buckets that needs to be sorted out for the downloads to work at all.

But if we fix CORS I'm hopeful that we can access the parquet files of the bq dump and get everything we need to make the full gcs or s3 path from that.

**Comment by s-paquette** [comment]
Ah I'd not thought about the CORS issue but yes that's ALSO something to deal with; nearly all the examples of this I've seen the origin is the same. 

The easiest solution might be to make an API endpoint that has parsed the parquet file (or a similar JSON structure) and then can simply be asked 'give me the URLs for this case/study/series' and it shoots them back.

**Comment by fedorov** [comment]
Feedback from @ulrikew: "Reporting size for a series download might be important for the SM files. I downloaded an SM series and the number of files that still needed to be downloaded did not change for a long while and I was wondering if the process had been disrupted. I went to the folder in which I placed the files and saw that two files were still pending and that gave me assurance that the process was working."

Suzanne says it is not difficult to update the downloaded size once series download is complete, which would be very helpful.

**Comment by s-paquette** [comment]
The message box now indicates a running total of the bytes requested, along with a breakout of collection, case, study, and series counts. Entire cases and studies can be downloaded as well.

@pieper @ulrikew What do we think the desired behavior should be if a user tries to download something which is already there? It's possible that they cancelled a download, or it was interrupted, etc. Should we warn them and cancel the download, and tell them to delete/rename any conflicting files before trying again? Thoughts?

**Comment by fedorov** [comment]
> Regarding re-downloads I think just overwriting is fine. 

I agree - I thought the same!

**Comment by s-paquette** [comment]
@fedorov @pgundluru @pieper Current version including above changes are now on dev and test tiers.

**Comment by pieper** [comment]
Thanks @s-paquette - the progress dialog looks nicer now but when I tested I got this issue:

<img width="1366" height="456" alt="Image" src="https://github.com/user-attachments/assets/39d03c76-592f-4933-b53b-b99f3d83e95c" />

This happens after I've already picked the directory so maybe something is being triggered as a side effect and it blocks the rest of the download.  This happens consistently for me when trying to download.

**Comment by s-paquette** [comment]
@pieper Can you give me the specific series that happened with? That error actually happens every time for reasons I've not yet been able to pin down but I've not yet had it stop a download.

**Comment by pieper** [comment]
@s-paquette that happens for me when I pick the first subject from B-mode-and-CEUS-Liver.

I'm trying a 4D-Lung case now and it seems to be working.

Another issue though with the status reporting is that the units seem to be off until it gets to 1GB.  You can see the dialog reports  a small number of MB, and the download seems very slow.  But the network tab reports 10x more or 100X more until it gets over a GB and then they are in sync and match what I get locally with `du`.  Also when it gets to GB it's still only showing 2 digits after the decimal so it feels slower so switching to 3 digits would be good.

<img width="1349" height="521" alt="Image" src="https://github.com/user-attachments/assets/0b76c6fe-f2c2-42dd-8856-9207af26883e" />

**Comment by s-paquette** [comment]
@pieper So, a couple of things. It can only report after a file has been fetched, because I don't have file size information until the response has been processed; it's not doing an as-it-streams-chunks calculation, but a 'completed' calculation. The network tab is showing you the streamed amounts. 

I can bump it out to 3 significant digits on GB but then it might look odd when it swaps over from MB--think that'll be alright?

**Comment by s-paquette** [comment]
@pieper @fedorov Okay, figured out the problem with that spurious error. it's the 'whole row click' option.

In order to allow the whole row to be a valid click space to open the lower levels of hierarchy, we have to put an event on it, and then have it ignore any clicks on specific location (eg. OHIF viewer, download button) and let other event bindings catch those as they bubble down. The problem is, that's not valid when using the file system API; the handler on the event which is initially triggered MUST be the one to call the file system API. If you try to let a second handler deal with it, you get this error and, as Steve as seen, some unexpected behavior. 

I'm trying to sort out a way to deal with this which doesn't involve a reworking of everything, but the whole-row-click adds in a significant level of complexity to event handling on these row buttons. One quick possibility is I can have clicking on the download button open a dialogue which in turn lets you choose your directory and confirm the download. Otherwise I need to redo how row clicks are managed to deal with this.

**Comment by pieper** [comment]
The 3 digit progress is nicer I think and it works well now for MB.  But when it gets to GB, it's still 3 digits but the last digit is always zero.

Regarding the row click thing, can the button be outside the row that's treated as one?  You don't really want to select the row when clicking download.

**Comment by s-paquette** [comment]
@pieper @fedorov Okay I think I have this sorted. It's going to update the display there every second (every 2 seconds felt too slow, every chunk was much too fast); I can adjust that depending on what we want, just let me know.  UI changes are slow and expensive so I'm trying not to do them too much. 

The download time estimate will go up or down depending on current transfer rate, so that's something to keep in mind. This is live on test.

**Comment by s-paquette** [comment]
@pieper @fedorov Alrighty, sped up the reporting a bit--note, if you're downloading a case or study with a large number of small files it will seem much slower due to the sheer number of round trip requests needed in comparison to a single large file. I've seen as slow as 100k/s on cases with 1500 files. On a single large file the download proceeds much more quickly.

**Comment by pieper** [comment]
I have no other concerns, I think we should go ahead and then revisit if anything pops up in production.  Thank you @s-paquette for all the work on this!

I agree there's nothing we can really do about the disk space issue.  I guess we could try saving to a small disk and see what the error condition is and make sure it's handled okay.
